### PR TITLE
= actor add cache to findConstructor

### PR DESCRIPTION
**Pull request by He-Pin**

### CODE REVIEW

File: akka-actor/src/main/scala/akka/util/Reflect.scala [hunk 1]
```diff
@@ -81,29 +87,55 @@ private[akka] object Reflect {
...
+      args.map({
+        case null    ⇒ "Any"
+        case notNull ⇒ notNull.getClass
+      }).mkString(",")
+    })"
+
+    @tailrec
+    def getConstructor(constructor: Constructor[T]): Constructor[T] = {
+      val constructorBinding = constructorBindingRef.get()
+      constructorBinding.get(constructorKey) match {
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/util/Reflect.scala]: Creation of that String key looks costly. It should be a case class with all parts separate.

**Review by patriknw — CHANGES_REQUESTED** [review]
What is the background for this performance optimization? Is this a real problem? I think we mostly promote the `Props(new MyActor(arg1, arg2))` pattern over `Props(MyActor.class, arg1, arg2)` and then it shouldn't matter?

**Comment by hepin1989** [comment]
Yes ,it is for `Props(MyActor.class, arg1, arg2)`.

File: akka-actor/src/main/scala/akka/util/Reflect.scala [hunk 2]
```diff
@@ -81,29 +89,53 @@ private[akka] object Reflect {
...
+            else cstrtr
+          } else null
+        }
+
+      if (constructor == null) error("no matching constructor")
+      else constructor
+    }
+
+    val constructorCacheKey = ConstructorCacheKey(clazz, args.size, args.collect({
+      case null    ⇒ classOf[Any]
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/util/Reflect.scala]: null, then I guess it should be `AnyRef`

File: akka-actor/src/main/scala/akka/util/Reflect.scala [hunk 3]
```diff
@@ -81,29 +89,53 @@ private[akka] object Reflect {
...
+            if (candidates.hasNext) error("multiple matching constructors")
+            else cstrtr
+          } else null
+        }
+
+      if (constructor == null) error("no matching constructor")
+      else constructor
+    }
+
+    val constructorCacheKey = ConstructorCacheKey(clazz, args.size, args.collect({
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/util/Reflect.scala]: args.size not needed

map instead of collect

**Review by patriknw — COMMENTED** [review]
perhaps adjust https://github.com/akka/akka/blob/master/akka-bench-jmh/src/main/scala/akka/actor/ActorCreationBenchmark.scala and see what improvement this gives

**Comment by hepin1989** [comment]
@patriknw rebased

**Comment by ktoso** [comment]
But... what difference does it make?

File: akka-bench-jmh/src/main/scala/akka/actor/ActorCreationWithReflectionCacheBenchmark.scala [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import scala.concurrent.Await
+import scala.concurrent.duration._
+
+/*
+without caching:
+[info] ActorCreationWithReflectionCacheBenchmark.synchronousStarting    ss    3  243.775 ± 330.837  us/op
+
+with caching:
+[info] ActorCreationWithReflectionCacheBenchmark.synchronousStarting    ss    3  482.001 ± 1494.089  us/op
+
```
Comment (ktoso) [inline akka-bench-jmh/src/main/scala/akka/actor/ActorCreationWithReflectionCacheBenchmark.scala]: us/op with caching is larger and bigger variance... are you sure this is improving things?

**Comment by hepin1989** [comment]
jmh:run -i 100 -wi 100 -f 1 -t 1 .*ActorCreationWithReflectionCacheBenchmark.*

**Comment by hepin1989** [comment]
```
[info] Result "akka.actor.ActorCreationWithReflectionCacheBenchmark.synchronousStarting":
[info]   N = 100
[info]   mean =    166.141 ±(99.9%) 10.784 us/op
[info]   Histogram, us/op:
[info]     [100.000, 125.000) = 1 
[info]     [125.000, 150.000) = 29 
[info]     [150.000, 175.000) = 43 
[info]     [175.000, 200.000) = 15 
[info]     [200.000, 225.000) = 7 
[info]     [225.000, 250.000) = 3 
[info]     [250.000, 275.000) = 1 
[info]     [275.000, 300.000) = 0 
[info]     [300.000, 325.000) = 0 
[info]     [325.000, 350.000) = 0 
[info]     [350.000, 375.000) = 1 
[info]   Percentiles, us/op:
[info]       p(0.0000) =    122.142 us/op
[info]      p(50.0000) =    157.605 us/op
[info]      p(90.0000) =    205.207 us/op
[info]      p(95.0000) =    225.325 us/op
[info]      p(99.0000) =    352.372 us/op
[info]      p(99.9000) =    353.308 us/op
[info]      p(99.9900) =    353.308 us/op
[info]      p(99.9990) =    353.308 us/op
[info]      p(99.9999) =    353.308 us/op
[info]     p(100.0000) =    353.308 us/op
[info] # Run complete. Total time: 00:00:01
[info] Benchmark                                                      Mode  Cnt    Score    Error  Units
[info] ActorCreationWithReflectionCacheBenchmark.synchronousStarting    ss  100  166.141 ± 10.784  us/op
```

File: akka-bench-jmh/src/main/scala/akka/actor/ActorCreationWithReflectionCacheBenchmark.scala (hunk 4 above)
Comment (hepin1989) [reply akka-bench-jmh/src/main/scala/akka/actor/ActorCreationWithReflectionCacheBenchmark.scala]: Sorry I was doing other things when I am running the bench. I read through the doc of sbt-jmh and re-run it.

**Comment by hepin1989** [comment]
Currently, it's the best I can get.

**Comment by hepin1989** [comment]
I have updated the number :)

File: akka-bench-jmh/src/main/scala/akka/actor/ActorCreationWithReflectionCacheBenchmark.scala [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    system.terminate()
+    Await.ready(system.whenTerminated, 15.seconds)
+  }
+
+  var current = 0
+  @Benchmark
+  @OutputTimeUnit(TimeUnit.MICROSECONDS)
+  def synchronousStarting =
+    system.actorOf(Props(classOf[MyNamedActor], 1, "2"), name)
+}
```
Comment (ktoso) [inline akka-bench-jmh/src/main/scala/akka/actor/ActorCreationWithReflectionCacheBenchmark.scala]: You should compare it to `Props(new MyNamedActor(1, "2"))` as that's the both a) fast and b) recommended way in any case.

I remain unconvienced this additional complexity is needed.

**Comment by patriknw** [comment]
I think we'll close this until there are some real evidence that this is important, since there are better ways of creating Props without using reflection.

**Pull request closed without merging**